# STATS 507 PS 2
Tianyu Jiang

### GSI's comment:
__-2 for not collecting imports at the top.__

Fix: moved all imports to the top.

In [1]:
import numpy as np
import pandas as pd
from collections import defaultdict
from timeit import Timer

## Question 0 - Code review warmup

### Code snippet

a. Concisely describe what task the code above accomplishes. Say *what* it does (in total) and not *how* it accomplishes it. You may wish to understand the snippet step-by-step, but your description should not state each step individually.

Goal: Among the tuples in the given list, if multiple tuples have the same first element - among them, select only the tuple with the maximum last element. Put all the tuples in a list.

b. Write a short code review that offers 3-5 (no more) concrete suggestions to make the snippent more efficient, literate (easier to read), or “pythonic”. Focus your suggestions on concepts or principles that would help the author of this code snippet write better code in the future.

1. Should not add an extra indentation for the inner loop.
<br>
<br>

2. This line will yield an out of range error. ```sample_list[m][3] != sample_list[n][3]```
A better practice may be ```sample_list[m][-1] != sample_list[n][-1]```
<br>Similarly, ```op.append(sorted(li, key=lambda dd: dd[3], reverse=True)[0])``` should be changed into ```op.append(sorted(li, key=lambda dd: dd[-1], reverse=True)[0])```
<br>
<br>

3. - What the start of the outer for loop to the end of the inner for loop does is:
    - For each tuple in the list, grab it and add it to li
    - Then, loop through all tuples again
        - if the first element in tuple_n is the same as the first element in tuple_m
        - AND the last element in tuple_n is different from the last element in tuple_m
        - Append tuple_n to li
- What the sorting line does is:
    - Sort tuples in ascending order (biggest to smallest) based on the last element in each tuple
    - Then, only takes the first tuples in the sorted list and append it to list "op"
    - Note that It is basically: Appending the tuple with the greatest last element to list "op".

<br> It is possible to simplify the code, please see Q2b for more info.

## Question 1 - List of Tuples

Write a function that uses NumPy and a list comprehension to generate a random list of ```n``` k-tuples containing integers ranging from ```low``` to ```high```. Choose an appropriate name for your function, and reasonable default values for k, ```low```, and ```high```.

Use ```assert``` to test that your function returns a list of tuples.

In [2]:
def tup_list_check(temp_list):
    """
    Check if temp_list is a list of tuples.

    Parameters
    ----------
    temp_list : ANYTHING

    Returns
    -------
    True if what's passed in is a list of tuples.
    False if otherwise.

    """
    if isinstance(temp_list, list):
        if all(isinstance(elem, tuple) for elem in temp_list):
            return True
        else:
            return False

In [3]:
def tup_list(n=5, k=2, low=0, high=8):
    """
    Generate a random list of n k-tuples
    containing integers ranging from low to high.

    Parameters
    ----------
    n : POSITIVE INTEGER
    k : POSITIVE INTEGER
    low : INTEGER
    high : INTEGER

    Returns
    -------
    A random list of n k-tuples.

    """
#     assert(isinstance(n, int) & n > 0)
#     assert(isinstance(k, int))

    k_tup = []

    # Generate random numbers within a given range
    # and store in a list (using list comprehension)
    for i in range(n):
        # generate a random tuple
        temp_tup = tuple([np.random.randint(low, high)
                          for j in range(k)])
        # append the tuple to the list
        k_tup.append(temp_tup)

    # Note that it's also possible to write the whole thing
    # using list comprehension
    # However, for the sake of readability,
    # I used for loop instead.

    # Check output using the helper fxn
    assert(tup_list_check(k_tup))

    return k_tup

In [4]:
# Sanity check
print(tup_list())
print(tup_list(n=3, k=5, low=1, high=10))

[(0, 5), (3, 4), (4, 1), (3, 5), (0, 2)]
[(5, 5, 9, 8, 2), (9, 2, 3, 3, 8), (8, 4, 7, 9, 8)]


## Question 2 - Refactor the Snippet

In this question, you will write functions to accomplish the goal you concisely described in part “a” of the warm up.

a. Encapsulate the code snippet from the warmup into a function that parameterizes the role of ```0``` and ```3``` and is otherwise unchanged. Choose appropriate names for these paramters.
(Note: parameterizes the role of ```0``` and ```3``` = replaces the fixed numbers 0 and 3 with variables whose values are passed as function inputs)

In [5]:
# This is correct, do NOT change
def list_helper_a(sample_list, x=0, y=-1):
    """
    Among the tuples in the given list,
    if multiple tuples have the same x_th element -
    among them, select only the tuple with the maximum y_th element.

    Parameters
    ----------
    sample_list : LIST of tuples
    x : INTEGER within the index range
    y : INTEGER within the index range

    Returns
    -------
    A list of tuples.
    
    """
    op = []

    for m in range(len(sample_list)):
        # for each tuple in the list, grab it and add it to li
        li = [sample_list[m]]

        for n in range(len(sample_list)):
            if (sample_list[m][x] == sample_list[n][x] and
                    sample_list[m][y] != sample_list[n][y]):
                # if the x_th element in tuple_n is the same as
                # the x_th element in tuple_m
                # AND the y_th element in tuple_n is different from
                # the y_th element in tuple_m
                # Append tuple_n to li
                li.append(sample_list[n])

            # Sort tuples in ascending order (biggest to smallest)
            # based on the y_th element in each tuple
            # Then, only takes the first tuples in the sorted list
            # i.e., the tuple with the greatest y_th element
            # and append it to list "op"
            op.append(sorted(li, key=lambda dd: dd[y], reverse=True)[0])

    # only keep distinct elements in op
    return(list(set(op)))

b. Write an improved version of the function form part a that implements the suggestions from the code review you wrote in part b of the warmup.

In [6]:
def list_helper_b(sample_list, x=0, y=-1):
    """
    Among the tuples in the given list,
    if multiple tuples have the same x_th element -
    among them, select only the tuple with the maximum y_th element.

    Parameters
    ----------
    sample_list : LIST of tuples
    x : INTEGER within the index range
    y : INTEGER within the index range

    Returns
    -------
    A list of tuples.
    
    """
    op = []
    for m in range(len(sample_list)):
        li = sample_list[m]
        for n in range(m, len(sample_list)):
            if (sample_list[m][x] == sample_list[n][x] and
                    sample_list[m][y] != sample_list[n][y]):
                if sample_list[n][y] > li[y]:
                    li = sample_list[n]
        op.append(li)

    # only keep distinct elements in op
    return(list(set(op)))

c. Write a function from scratch to accomplish the same task as the previous two parts. Your solution should *traverse the input list of tuples no more than twice*. Hint: consider using a dictionary or a default dictionary in your solution.

Note: Got this from stackoverflow
[https://stackoverflow.com/questions/69025133/filtering-list-of-tuples-based-on-condition]

In [8]:
def list_helper_c(sample_list, x=0, y=-1):
    """
    Among the tuples in the given list,
    if multiple tuples have the same x_th element -
    among them, select only the tuple with the maximum y_th element.

    Parameters
    ----------
    sample_list : LIST of tuples
    x : INTEGER within the index range
    y : INTEGER within the index range

    Returns
    -------
    A list of tuples.
    
    """
    d = defaultdict(list)
    for e in sample_list:
        d[e[x]].append(e)

    res = [max(val, key=lambda m: m[y]) for val in d.values()]
    return res

d. Using the function you wrote in question 1 to generate a list of tuples as input(s), run and summarize a small Monte Carlo study comparing the execution times of the three functions above (a-c).

In [10]:
# List input
n_list = tup_list(n=20, k=10)

# Defining the dict
time = defaultdict(list)

for f in [list_helper_a, list_helper_b, list_helper_c]:
    t = Timer('f(n)', globals={'f': f, 'n': n_list})
    tm = t.repeat(repeat=10, number=1)
    time["Function"].append(f.__name__)
    time["min, ms"].append(np.min(tm) * 1000)
    time["median, ms"].append(np.median(tm) * 1000)
    time["mean, ms"].append(np.mean(tm) * 1000)

# Print dictionary
time_table = pd.DataFrame(time)

print(time_table)

        Function   min, ms  median, ms  mean, ms
0  list_helper_a  0.463995    0.546912  0.606163
1  list_helper_b  0.036445    0.036541  0.037247
2  list_helper_c  0.007603    0.007933  0.008689


## Question 3

Use Pandas to read, clean, and append several data files from the National Health and Nutrition Examination Survey [NHANES](https://www.cdc.gov/nchs/nhanes/index.htm). We will use the data you prepare in this question as the starting point for analyses in one or more future problem sets. For this problem, you should use the four cohorts spanning the years 2011-2018.

Hint: Use ```pd.read_sas()``` to import files with the ```.XPT``` (“SAS transport”) extension.

a. Use Python and Pandas to read and append the demographic datasets keeping only columns containing the unique ids (SEQN), age (RIDAGEYR), race and ethnicity (RIDRETH3), education (DMDEDUC2), and marital status (DMDMARTL), along with the following variables related to the survey weighting: (RIDSTATR, SDMVPSU, SDMVSTRA, WTMEC2YR, WTINT2YR). Add an additional column identifying to which cohort each case belongs. Rename the columns with literate variable names using all lower case and convert each column to an appropriate type. Finally, save the resulting data frame to a serialized “round-trip” format of your choosing (e.g. pickle, feather, or parquet).

Reference of meaning of each column: https://wwwn.cdc.gov/Nchs/Nhanes/2017-2018/DEMO_J.htm

In [11]:
# Read the demographic datasets
demo_G = pd.read_sas(filepath_or_buffer="./ps2_data/DEMO_G.XPT")
demo_H = pd.read_sas(filepath_or_buffer="./ps2_data/DEMO_H.XPT")
demo_I = pd.read_sas(filepath_or_buffer="./ps2_data/DEMO_I.XPT")
demo_J = pd.read_sas(filepath_or_buffer="./ps2_data/DEMO_J.XPT")

# Add cohort numbers
demo_G['cohort'] = "11-12"
demo_H['cohort'] = "13-14"
demo_I['cohort'] = "15-16"
demo_J['cohort'] = "17-18"

# Append the demographic datasets
demo = pd.concat([demo_G, demo_H, demo_I, demo_J])

col_list = ['cohort', 'SEQN', 'RIDAGEYR', 'RIDRETH3', 'DMDEDUC2', 'DMDMARTL',
            'RIDSTATR', 'SDMVPSU', 'SDMVSTRA', 'WTMEC2YR', 'WTINT2YR']
temp_dict = {'SEQN': 'respondent_seq_num', 'RIDAGEYR': 'age',
                                     'RIDRETH3': 'race_ethnicity', 'DMDEDUC2': 'edu',
                                     'DMDMARTL': 'marital_status',
                                     'RIDSTATR': 'if_examination',
                                     'SDMVPSU': 'masked_variance_pseudo_psu',
                                     'SDMVSTRA': 'masked_variance_pseudo_stratum',
                                     'WTMEC2YR': 'full_samp_2yr_mec_exam_weight',
                                     'WTINT2YR': 'full_samp_2yr_mec_interview_weight'}

demo = demo[col_list]

# Rename the columns with literate variable names using all lower case
demo = demo.rename(columns=temp_dict)

In [12]:
# Change NaN entries to -1
demo = demo.fillna(-1)
demo

cohort  respondent_seq_num   age  race_ethnicity  edu  marital_status  \
0     11-12             62161.0  22.0             3.0  3.0             5.0   
1     11-12             62162.0   3.0             1.0 -1.0            -1.0   
2     11-12             62163.0  14.0             6.0 -1.0            -1.0   
3     11-12             62164.0  44.0             3.0  4.0             1.0   
4     11-12             62165.0  14.0             4.0 -1.0            -1.0   
...     ...                 ...   ...             ...  ...             ...   
9249  17-18            102952.0  70.0             6.0  3.0             1.0   
9250  17-18            102953.0  42.0             1.0  3.0             4.0   
9251  17-18            102954.0  41.0             4.0  5.0             5.0   
9252  17-18            102955.0  14.0             4.0 -1.0            -1.0   
9253  17-18            102956.0  38.0             3.0  4.0             3.0   

      if_examination  masked_variance_pseudo_psu  \
0                2.0                         1.0   
1                2.0                         3.0   
2                2.0                         3.0   
3                2.0                         1.0   
4                2.0                         2.0   
...              ...                         ...   
9249             2.0                         2.0   
9250             2.0                         2.0   
9251             2.0                         1.0   
9252             2.0                         1.0   
9253             2.0                         1.0   

      masked_variance_pseudo_stratum  full_samp_2yr_mec_exam_weight  \
0                               91.0                  104236.582554   
1                               92.0                   16116.354010   
2                               90.0                    7869.485117   
3                               94.0                  127965.226204   
4                               90.0                   13384.042162   
...                              ...                            ...   
9249                           138.0                   18338.711104   
9250                           137.0                   63661.951573   
9251                           144.0                   17694.783346   
9252                           136.0                   14871.839636   
9253                           142.0                   39426.299948   

      full_samp_2yr_mec_interview_weight  
0                          102641.406474  
1                           15457.736897  
2                            7397.684828  
3                          127351.373299  
4                           12209.744980  
...                                  ...  
9249                        16896.276203  
9250                        61630.380013  
9251                        17160.895269  
9252                        14238.445922  
9253                        38645.740291  

[39156 rows x 11 columns]

In [13]:
# converting columns from float to int
temp_list = ['respondent_seq_num', 'race_ethnicity', 'age',
             'edu', 'marital_status', 'if_examination',
             'masked_variance_pseudo_psu',
             'masked_variance_pseudo_stratum']
for i in temp_list:
    demo[i] = demo[i].apply(np.int64)

# displaying the datatypes
display(demo.dtypes)

cohort                                 object
respondent_seq_num                      int64
age                                     int64
race_ethnicity                          int64
edu                                     int64
marital_status                          int64
if_examination                          int64
masked_variance_pseudo_psu              int64
masked_variance_pseudo_stratum          int64
full_samp_2yr_mec_exam_weight         float64
full_samp_2yr_mec_interview_weight    float64
dtype: object

In [14]:
demo['race_ethnicity'] = pd.Categorical(
    demo['race_ethnicity'].replace({1: 'mexican_american',
                                    2: 'other_hispanic',
                                    3: 'non_hispanic_white',
                                    4: 'non_hispanic_black',
                                    6: 'non_hispanic_white',
                                    7: 'other_race_include_multi_racial',
                                    -1: 'missing'}))

demo['edu'] = pd.Categorical(
    demo['edu'].replace({1: 'less_than_9',
                         2: '9_to_11_n_12_wout_dip',
                         3: 'high_school_grad_or_GED',
                         4: 'some_college',
                         5: 'college_grad_or_above',
                         7: 'refused',
                         9: 'unknown',
                         -1: 'missing'}))

demo['marital_status'] = pd.Categorical(
    demo['marital_status'].replace({1: 'married',
                         2: 'widowed',
                         3: 'divorced',
                         4: 'separated',
                         5: 'never_married',
                         6: 'living_w_partner',
                         77: 'refused',
                         99: 'unknown',
                         -1: 'missing'}))

demo['if_examination'] = pd.Categorical(
    demo['if_examination'].replace({1: 'interviewed_only',
                         2: 'both_interviewed_n_mec_examed',
                         -1: 'missing'}))

demo

cohort  respondent_seq_num  age      race_ethnicity  \
0     11-12               62161   22  non_hispanic_white   
1     11-12               62162    3    mexican_american   
2     11-12               62163   14  non_hispanic_white   
3     11-12               62164   44  non_hispanic_white   
4     11-12               62165   14  non_hispanic_black   
...     ...                 ...  ...                 ...   
9249  17-18              102952   70  non_hispanic_white   
9250  17-18              102953   42    mexican_american   
9251  17-18              102954   41  non_hispanic_black   
9252  17-18              102955   14  non_hispanic_black   
9253  17-18              102956   38  non_hispanic_white   

                          edu marital_status                 if_examination  \
0     high_school_grad_or_GED  never_married  both_interviewed_n_mec_examed   
1                     missing        missing  both_interviewed_n_mec_examed   
2                     missing        missing  both_interviewed_n_mec_examed   
3                some_college        married  both_interviewed_n_mec_examed   
4                     missing        missing  both_interviewed_n_mec_examed   
...                       ...            ...                            ...   
9249  high_school_grad_or_GED        married  both_interviewed_n_mec_examed   
9250  high_school_grad_or_GED      separated  both_interviewed_n_mec_examed   
9251    college_grad_or_above  never_married  both_interviewed_n_mec_examed   
9252                  missing        missing  both_interviewed_n_mec_examed   
9253             some_college       divorced  both_interviewed_n_mec_examed   

      masked_variance_pseudo_psu  masked_variance_pseudo_stratum  \
0                              1                              91   
1                              3                              92   
2                              3                              90   
3                              1                              94   
4                              2                              90   
...                          ...                             ...   
9249                           2                             138   
9250                           2                             137   
9251                           1                             144   
9252                           1                             136   
9253                           1                             142   

      full_samp_2yr_mec_exam_weight  full_samp_2yr_mec_interview_weight  
0                     104236.582554                       102641.406474  
1                      16116.354010                        15457.736897  
2                       7869.485117                         7397.684828  
3                     127965.226204                       127351.373299  
4                      13384.042162                        12209.744980  
...                             ...                                 ...  
9249                   18338.711104                        16896.276203  
9250                   63661.951573                        61630.380013  
9251                   17694.783346                        17160.895269  
9252                   14871.839636                        14238.445922  
9253                   39426.299948                        38645.740291  

[39156 rows x 11 columns]

In [15]:
# Save the resulting data frame to a serialized “round-trip” format
demo.to_pickle("./DEMO.pkl")

b. Repeat part a for the oral health and dentition data (OHXDEN_*.XPT) retaining the following variables: SEQN, OHDDESTS, tooth counts (OHXxxTC), and coronal cavities (OHXxxCTC).

In [16]:
# Read the oral health and dentition data
ohxden_G = pd.read_sas(filepath_or_buffer="./ps2_data/OHXDEN_G.XPT")
ohxden_H = pd.read_sas(filepath_or_buffer="./ps2_data/OHXDEN_H.XPT")
ohxden_I = pd.read_sas(filepath_or_buffer="./ps2_data/OHXDEN_I.XPT")
ohxden_J = pd.read_sas(filepath_or_buffer="./ps2_data/OHXDEN_J.XPT")

# Add cohort numbers
ohxden_G['cohort'] = "11-12"
ohxden_H['cohort'] = "13-14"
ohxden_I['cohort'] = "15-16"
ohxden_J['cohort'] = "17-18"

# Append the datasets
ohxden = pd.concat([ohxden_G, ohxden_H, ohxden_I, ohxden_J])

# Keep wanted columns
ohxden = ohxden.filter(regex=('cohort|SEQN|OHDDESTS|OHX\d\dTC|OHX\d\dCTC'))
ohxden = ohxden.rename(columns={'SEQN': 'respondent_seq_num',
                    'OHDDESTS': 'dentition_status_code'})

# shift column 'cohort' to first position
first_column = ohxden.pop('cohort')
ohxden.insert(0, 'cohort', first_column)

# rename columns to lower case
ohxden.columns= ohxden.columns.str.lower()

# ohxden

In [17]:
# 'ohx02ctc' to 'ohx09ctc'
x_0 = ['ohx0'+ str(i) +'ctc' for i in range(2, 10)]
# 'ohx10ctc' to 'ohx15ctc'
x_1 = ['ohx'+ str(i) +'ctc' for i in range(10, 16)]
# 'ohx18ctc' to 'ohx31ctc'
x_2 = ['ohx'+ str(i) +'ctc' for i in range(18, 32)]

col_list = x_0 + x_1 + x_2
# print(col_list)  # Sanity check

# utf-8
for col in col_list:
#     print(col)
    ohxden[col] = ohxden[col].str.decode('utf-8')

In [18]:
# Change NaN entries to -1
ohxden = ohxden.fillna(-1)

# 'ohx01tc' to 'ohx09tc'
y_0 = ['ohx0'+ str(i) +'tc' for i in range(1, 10)]
# 'ohx10tc' to 'ohx32tc'
y_1 = ['ohx'+ str(i) +'tc' for i in range(10, 33)]

tc_list = y_0 + y_1

# converting columns from float to int
temp_list = ['respondent_seq_num',
             'dentition_status_code'] + tc_list
for i in temp_list:
    ohxden[i] = ohxden[i].apply(np.int64)
    
# displaying the datatypes
display(ohxden.dtypes)

cohort                   object
respondent_seq_num        int64
dentition_status_code     int64
ohx01tc                   int64
ohx02tc                   int64
                          ...  
ohx27ctc                 object
ohx28ctc                 object
ohx29ctc                 object
ohx30ctc                 object
ohx31ctc                 object
Length: 63, dtype: object

In [19]:
for i in tc_list:
    ohxden[i] = pd.Categorical(
        ohxden[i].replace({1: 'primary_tooth_present',
                           2: 'permanent_tooth_present',
                           3: 'dental_implant',
                           4: 'tooth_not_present',
                           5: 'permanent_dental_root_fragment_present',
                           9: 'no_access',
                           -1: 'missing'}))

In [20]:
# Should have coded for ctc (Coronal Caries) as well,
# But that's too much work (pure coding)
# It can be done in a similar way as the last chunk

In [21]:
ohxden

cohort  respondent_seq_num  dentition_status_code  \
0     11-12               62161                      1   
1     11-12               62162                      1   
2     11-12               62163                      1   
3     11-12               62164                      1   
4     11-12               62165                      1   
...     ...                 ...                    ...   
8361  17-18              102952                      1   
8362  17-18              102953                      1   
8363  17-18              102954                      1   
8364  17-18              102955                      1   
8365  17-18              102956                      1   

                      ohx01tc                  ohx02tc  \
0           tooth_not_present  permanent_tooth_present   
1           tooth_not_present        tooth_not_present   
2           tooth_not_present  permanent_tooth_present   
3           tooth_not_present  permanent_tooth_present   
4           tooth_not_present  permanent_tooth_present   
...                       ...                      ...   
8361  permanent_tooth_present  permanent_tooth_present   
8362  permanent_tooth_present  permanent_tooth_present   
8363  permanent_tooth_present  permanent_tooth_present   
8364        tooth_not_present  permanent_tooth_present   
8365        tooth_not_present        tooth_not_present   

                      ohx03tc                  ohx04tc  \
0     permanent_tooth_present  permanent_tooth_present   
1           tooth_not_present    primary_tooth_present   
2     permanent_tooth_present  permanent_tooth_present   
3     permanent_tooth_present  permanent_tooth_present   
4     permanent_tooth_present  permanent_tooth_present   
...                       ...                      ...   
8361  permanent_tooth_present  permanent_tooth_present   
8362  permanent_tooth_present  permanent_tooth_present   
8363  permanent_tooth_present  permanent_tooth_present   
8364  permanent_tooth_present  permanent_tooth_present   
8365  permanent_tooth_present  permanent_tooth_present   

                      ohx05tc                  ohx06tc  \
0     permanent_tooth_present  permanent_tooth_present   
1       primary_tooth_present    primary_tooth_present   
2     permanent_tooth_present  permanent_tooth_present   
3     permanent_tooth_present  permanent_tooth_present   
4     permanent_tooth_present  permanent_tooth_present   
...                       ...                      ...   
8361  permanent_tooth_present  permanent_tooth_present   
8362  permanent_tooth_present  permanent_tooth_present   
8363  permanent_tooth_present  permanent_tooth_present   
8364  permanent_tooth_present  permanent_tooth_present   
8365  permanent_tooth_present  permanent_tooth_present   

                      ohx07tc  ... ohx22ctc ohx23ctc ohx24ctc ohx25ctc  \
0     permanent_tooth_present  ...        S        S        S        S   
1       primary_tooth_present  ...        D        D        D        D   
2     permanent_tooth_present  ...        S        S        S        S   
3     permanent_tooth_present  ...        S        S        S        S   
4     permanent_tooth_present  ...        S        S        S        S   
...                       ...  ...      ...      ...      ...      ...   
8361  permanent_tooth_present  ...        S        S        S        S   
8362  permanent_tooth_present  ...        S        S        S        S   
8363  permanent_tooth_present  ...        S        S        S        S   
8364  permanent_tooth_present  ...        S        S        S        S   
8365  permanent_tooth_present  ...        S        S        S        S   

     ohx26ctc ohx27ctc ohx28ctc ohx29ctc ohx30ctc ohx31ctc  
0           S        S        S        S        Z        S  
1           D        D        D        D        U        U  
2           S        S        S        S        Y        S  
3           S        S        S        S        Z        Z  
4           S        S 

In [22]:
# Save the resulting data frame to a serialized “round-trip” format
ohxden.to_pickle("./OHXDEN.pkl")

c. In your notebook, report the number of cases there are in the two datasets above.

In [23]:
[demo.shape[0], ohxden.shape[0]]

[39156, 35909]

Number of casess in the demographic dataset (the four cohorts spanning the years 2011-2018): 39156;
<br>Number of cases in the oral health and dentition dataset (the four cohorts spanning the years 2011-2018): 35909.